### ISSUE TITLE
[1.12-rc2] daemon.ID should be replaced with swarm NodeID

### ISSUE THREAD

**Issue by stevvooe** [moby/moby#24095]
Currently, the output of `docker info` and the `Daemon` struct have an identifier set with the `trustKey`. Since this isn't really used, we should probably set this to the actual cryptographic node identity when the daemon changes to swarm mode.

This will ensure that we have a really node identity associated with an engine and ensure that it is available to users.

See docker/swarmkit#1075

cc @diogomonica @dmcgowan @aluzzardi

**Comment by yongtang** [comment]
@stevvooe I checked the `docker info`, it seems that if daemon is in swarm mode, then the output will consists of a `Swarm` section with `NodeID` and `IsManager` field:
```
...
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: null host bridge overlay
Swarm: active
 NodeID: 13cjpir2ahv0ie0e12lwyghl7
 IsManager: No
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-24-generic
Operating System: Debian GNU/Linux 8 (jessie) (containerized)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.788 GiB
Name: 6d1de31cb302
ID: YGEZ:MICB:ZC6T:K2BL:X4IH:5QWC:TEKE:2Y44:IFKJ:ADAJ:QJKP:NML6
...
```
Maybe we could move the `NodeID` to top level and co-exist with the `ID`? As a daemon may join or leave the swarm, maybe we want to keep both `NodeID` and `ID`?

**Comment by diogomonica** [comment]
A node should probably only have an ID when it is in swarm mode.

**Comment by stevvooe** [comment]
@diogomonica Agreed, but we should probably deprecate daemon.ID, as that is fairly useless.

**Comment by dmcgowan** [comment]
The trust key is still used to push schema 1 manifests. I am fine getting rid of the trust key as long as we have private key to replace it with. If a new key is only generated when swarm mode is activated then we can't get rid of the trust key. If we can generate it on first startup like we do with the trust key then I don't have a problem just importing that to use for the v1 manifests and daemon ID.

**Comment by stevvooe** [comment]
@dmcgowan We can keep the trust key, but do we have to rely on that as the source of daemon identity?

**Comment by dmcgowan** [comment]
@stevvooe absolutely not, it was originally used for swarm since it was reliably unique per daemon and always generated at startup. It can be replaced with any random identifier since its cryptographic signature is never checked. It was useful for checking if a daemon signed a schema1 manifest but there was no tooling around that and schema1 manifest is now deprecated.

**Comment by thaJeztah** [comment]
ping @diogomonica @dmcgowan what's the status on this one?

**Comment by diogomonica** [comment]
Should probably be done. The assumption is that no one is relying on the current ID to have the current format, and that we can simply replace it by the swarm ID.

**Comment by stevvooe** [comment]
@dmcgowan So, are we comfortable with the possibility that this node ID could change during the lifetime of the engine? For example, if a node is not part of a swarm and joins, it will get a new ID. If node joins and leaves, it will get a new id.

I think we should link these concepts, but is this a good idea overall?

**Comment by dmcgowan** [comment]
I agree with Diogo that it is safe to assume the format is not being relied on, but it is possible that tooling is expecting this ID to not mutate. Could this cause problems with updating from swarm v1 to swarm mode?

I agree that basing it off the trust key is no longer useful, but why not keep the ID specific to the machine and use the Swarm NodeID when it is needed. Are we going to define some method to verify the cryptographic ID of a node, if not all we need is a UUID. The only reason we used a cryptographic key before is we had plans for using it for signing and a known hosts feature.

**Comment by stevvooe** [comment]
> I agree with Diogo that it is safe to assume the format is not being relied on, but it is possible that tooling is expecting this ID to not mutate. Could this cause problems with updating from swarm v1 to swarm mode?

Unless it is specified that this is a UUID, reliance on the format of an opaque field shouldn't be considered supported.

> Are we going to define some method to verify the cryptographic ID of a node, if not all we need is a UUID.

We already effectively do this with swarm. We should probably extend this functionality as a property of the nodes identity. This is a powerful concept.

**Comment by thaJeztah** [comment]
We removed the use of libtrust for this ID, and instead use an ID that's stored in `/var/lib/docker/engine-id` (see https://github.com/moby/moby/pull/44329).

For existing installs, we copied the libtrust ID, and for new installs (or the `engine-id` not being present), we write a UUID to the file, but the format of the ID is not formalised (users can write their own ID to this file if they want their own form of identification).

w.r.t. using the Swarm Node ID for this purpose; we _could_ write that ID to the file, but I think it's ok to keep them separate; `docker info` _does_  provide the node-ID in the output though, so users are able to find the node ID from both manager and worker nodes if they want to use the _swarm_ node ID for identification;

Manager node:

```
Swarm: active
  NodeID: 2lm9w9kbepgvkzkkeyku40e65
  Is Manager: true
  ClusterID: abajmipo7b4xz5ip2nrla6b11
```

Worker node:

```
Swarm: active
  NodeID: n41b2cijmhifxxvz56vwrs12q
  Is Manager: false
```

I'll close this ticket.
